### PR TITLE
[DeadCode] Handle anonymous and arrow function uses in RemoveJustPropertyFetchRector

### LINKED ISSUE #7263: Internal error in Rector\DeadCode\Rector\StmtsAwareInterface\RemoveJustPropertyFetchRector
# Bug Report

| Subject        | Details  |
| :------------- | :-----------|
| Rector version | v0.13.5 |

## Minimal PHP Code Causing Issue

https://getrector.org/demo/05195c1f-810d-4cc6-baf7-c9f130b49ebf

## Expected Behaviour

If the anonymous function is static - rector should skip it.
If the function is not static - variable can be inlined and `use` removed.
```php
<?php

final class DemoFile
{
    private SomeClass $someObj;

    public function run(array $array)
    {
        return array_map(function ($item) {
            return $this->someObj->someMethod($item);
        }, $array);
    }
}
```

### PR DESCRIPTION
**Pull request by kick-the-bucket**
Fixes rectorphp/rector#7263

### CODE REVIEW

File: rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php [hunk 1]
```diff
@@ -104,14 +105,22 @@ public function refactor(Node $node): ?Node
             }
 
             $followingStmts = array_slice($stmts, $key + 1);
+            if (count($followingStmts) > 0
+                && $followingStmts[0] instanceof Expression
+                && $followingStmts[0]->expr instanceof Node\Expr\Closure
+                && $followingStmts[0]->expr->static
```
Comment (samsonasik) [inline rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php]: you can move this to separate method, eg: `shouldSkip(): bool` to reduce complexity.

File: rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php [hunk 2]
```diff
@@ -147,14 +158,29 @@ private function replaceVariablesWithPropertyFetch(
         // remove assign node
         unset($stmtsAware->stmts[$currentStmtsKey]);
 
+        // avoid crash in \PHPStan\Analyser\MutatingScope::enterAnonymousFunctionWithoutReflection
+        $variableCompatibleFetch = clone $propertyFetch;
+        if (!$variableCompatibleFetch->name instanceof Node\Identifier && !is_string($variableCompatibleFetch->name)) {
+            throw new ShouldNotHappenException();
```
Comment (samsonasik) [inline rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php]: I think `return null` can be used instead, and make return type declaration nullable: `: ?StmtsAwareInterface {`

File: rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php [hunk 3]
```diff
@@ -147,14 +158,29 @@ private function replaceVariablesWithPropertyFetch(
...
+        }
+        $variableCompatibleFetch->name = (string) $variableCompatibleFetch->name;
```
Comment (samsonasik) [inline rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php]: cast `(string)` is not needed, replace with `Identifier` should be ok.

Comment (samsonasik) [reply rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php]: actually, it seems should can be removed, because assign to same data or Identifier.

Comment (kick-the-bucket) [reply rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php]: This is the key of the fix - converting `Identifier` to `string`, because PhpStan throws an exception if it gets a name that is not a string

Comment (kick-the-bucket) [reply rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php]: This is where it crashed before this fix:
https://github.com/phpstan/phpstan-src/blob/1.8.x/src/Analyser/MutatingScope.php#L2846

Comment (samsonasik) [reply rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php]: I tried remove the assignment and it seems working ok:

```diff
-        $variableCompatibleFetch->name = (string) $variableCompatibleFetch->name;
```

```bash
vendor/bin/phpunit rules-tests/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector/RemoveJustPropertyFetchRectorTest.php
PHPUnit 9.5.21 #StandWithUkraine

............                                                      12 / 12 (100%)

Time: 00:01.518, Memory: 96.50 MB

OK (12 tests, 24 assertions)
```

Comment (kick-the-bucket) [reply rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php]: Looks like with all the additional edge case handling this work-around is not needed anymore, will remove.

File: rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php [hunk 4]
```diff
@@ -143,18 +161,33 @@ private function replaceVariablesWithPropertyFetch(
         int $currentStmtsKey,
         array $variableUsages,
         PropertyFetch $propertyFetch
-    ): StmtsAwareInterface {
+    ): ?StmtsAwareInterface {
         // remove assign node
         unset($stmtsAware->stmts[$currentStmtsKey]);
 
+        // avoid crash in \PHPStan\Analyser\MutatingScope::enterAnonymousFunctionWithoutReflection
+        $variableCompatibleFetch = clone $propertyFetch;
```
Comment (samsonasik) [inline rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php]: clone `$propertyFetch` is not needed, just use the `$propertyFetch` variable.

File: phpstan.neon [hunk 5]
```diff
@@ -666,3 +666,8 @@ parameters:
         -
             message: '#New objects with "\$fullyQualifiedObjectType" name are overridden\. This can lead to unwanted bugs, please pick a different name to avoid it#'
             path: packages/NodeTypeResolver/PhpDocNodeVisitor/NameImportingPhpDocNodeVisitor.php
+
+        -
+            message: "#^Property PhpParser\\\\Node\\\\Expr\\\\PropertyFetch\\:\\:\\$name \\(PhpParser\\\\Node\\\\Expr\\|PhpParser\\\\Node\\\\Identifier\\) does not accept string\\.$#"
+            count: 1
+            path: rules/DeadCode/Rector/StmtsAwareInterface/RemoveJustPropertyFetchRector.php
```
Comment (samsonasik) [inline phpstan.neon]: this can be removed.

**Comment by samsonasik** [comment]
Thank you, could you update PR title to something like: 

```
[DeadCode] Handle indirect used in next ClosureUse_ on RemoveJustPropertyFetchRector
```

and PR description to:

```
Fixes https://github.com/rectorphp/rector/issues/7263
```

**Review by samsonasik — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thank you :+1: 

Great team work :wink:
